pkg/modules/notary/notary.go

```diff
@@ -0,0 +1,300 @@
+	// minNotValidBefore is the minimum NVB value among fallbacks transactions.
+	// We stop trying to send mainTx to the network if the chain reaches minNotValidBefore height.
+	minNotValidBefore uint32
+	fallbacks         []*transaction.Transaction
+	// nKeys is the number of witnesses to be collected for the main transaction
+	nKeys uint8
+	// nKeysCollected is the number of already collected witnesses
+	nKeysCollected uint8
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().Modules.P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	var isValid bool
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, err := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	if err == nil { // valid main tx signatures
+		isValid = true
+	}
+	r, ok := n.requests[payload.MainTransaction.Hash()]
+	if ok {
+		if r.nKeysCollected >= nKeys { // already collected sufficient number of keys to complete main transaction
+			return // TODO: however, we may still accept extra fallback transactions without adding new signatures to main tx
+		}
+		r.isValid = isValid && r.isValid && typ == r.typ // request for the main tx is found, but witnesses types may be different
+		if nvbFallback < r.minNotValidBefore {
+			r.minNotValidBefore = nvbFallback
+		}
+	} else {
+		r = request{
+			nKeys:             nKeys,
+			main:              payload.MainTransaction,
+			typ:               typ,
+			isValid:           isValid,
+			minNotValidBefore: nvbFallback,
+		}
+	}
+	r.fallbacks = append(r.fallbacks, payload.FallbackTransaction)
+	if r.isValid {
+		r.nKeysCollected++
+		for i, w := range payload.MainTransaction.Scripts {
+			if len(w.InvocationScript) != 0 {
+				switch r.typ {
+				case Signature:
+					r.main.Scripts[i] = w
+				case MultiSignature:
+					r.main.Scripts[i].InvocationScript = append(r.main.Scripts[i].InvocationScript, payload.MainTransaction.Scripts[i].InvocationScript...) // TODO: sort?
```

Comment:
You have to parse and sort (by public key), otherwise multisig won't work.

Reply:
Done, but it needs a review. Now we can have either multiple signature witnesses or single multisignature witness with one signature filled.